Forescout Host Property Monitor solution - initial version

fs-connect opened the pull request:
Change(s):
   - Initial version of solution Forescout Host Property Monitor 

   Reason for Change(s):
   - New Solution development for Forescout Host Property Monitor

   Version Updated:
   - No

   Testing Completed:
   - Yes

   Checked that the validations are passing and have addressed any issues that are present:
   - Yes

v-spadarthi commented:
@devikamehra /rushriva : Please review the Playbooks.

v-spadarthi commented:
@aprakash13 :Please review the analytic rules

v-spadarthi commented:
@fs-connect : Could you please resolve the below errors,
Please fix the validation errors Analytics Rules files extension should be  " .YAML" please see the below screenshot
<img width="457" alt="image" src="https://user-images.githubusercontent.com/101796244/182543291-b0ca5361-a8e6-405c-98cc-d1a7f48d4954.png">
While deploying CreateUIdefinition it should display all the content like  Playbooks and Analytics but, it is displaying Playbooks only Please see the below screenshot below,
<img width="476" alt="image" src="https://user-images.githubusercontent.com/101796244/182543695-2314f392-3d01-48b3-8a56-8c1616e71e1b.png">
While deploying Maintemplate we are getting errors please see the screenshot below,
<img width="952" alt="image" src="https://user-images.githubusercontent.com/101796244/182544124-20bc978d-40b9-4d06-883a-81332a73647c.png">

aprakash13 reviewed (changes requested):
The Analytic rule templates in Github follows a specific format. Here is a link to an example: https://github.com/Azure/Azure-Sentinel/blob/ae40b7e77fd406a5ea0297a7916c564eed2fc3c4/Detections/AWSGuardDuty/AWS_GuardDuty_template.yaml
Can we please have the query modified to the suggested format.

fs-connect commented:
Thanks for the review. I have updated files to address comments.
I request reviewers to review the new updates. Thanks!

v-spadarthi commented:
CreateUIDefinition and Maintemplate looks good.
<img width="449" alt="image" src="https://user-images.githubusercontent.com/101796244/183286885-1c9b2494-686c-414b-b718-2b8ebc2b1a66.png">
<img width="935" alt="image" src="https://user-images.githubusercontent.com/101796244/183286770-d6d29eb2-abc2-4c98-a338-c301d3192888.png">

@aprakash13 : Please review the analytic rules.
@devikamehra /rushriva : Please review the Playbooks.

aprakash13 reviewed (commented):
The query looks good. One small suggestion is to move the version information to the end of the query just before 'kind': 
![image](https://user-images.githubusercontent.com/37783395/184307184-fe3bc070-08d3-40b1-bdd6-a0b0fe7cb9e2.png)

Please feel free to merge if other components looks good.

v-spadarthi commented:
@devikamehra /rushriva : Please review the Playbooks.

devikamehra commented on Solutions/Forescout Host Property Monitor/Analytic Rules/DnsniffEvent-1.yaml:
```diff
@@ -0,0 +1,30 @@
+id: d272e277-f285-4dbc-ae2d-7f65ba64a79e
+name: DnsniffEvent-1
```
Can we give a suitable name for this Analytical Rule.

fs-connect replied:
Our customers are familiar with and understand the name DnsniffEvent. And we are planning to add more examples later, so we added a number to the name.

devikamehra replied:
I understand that your customers are aware of the termologies you are using. But we are keeping in mind the interest of all the users of Sentinel. Please update the names so that they reflect the operation to be performed.

fs-connect replied:
Understand, we will update the names.

devikamehra commented on Solutions/Forescout Host Property Monitor/Playbooks/DnsniffEventPlaybook-1.json:
```diff
@@ -0,0 +1,386 @@
+    "contentVersion":  "1.0.0.0",
+    "parameters":  {
+        "PlaybookName":  {
+            "defaultValue":  "DnsniffEventPlaybook-1",
```
Can we give a suitable name for the playbook file and playbook. This would help the users understand the operation of the playbook better.

fs-connect replied:
Our customers are familiar with and understand the name DnsniffEvent. And we are planning to add more examples later, so we added a number to the name.

fs-connect commented on .script/tests/KqlvalidationsTests/CustomTables/ForescoutHostProperties_CL.json:
```diff
@@ -0,0 +1,25 @@
+{
```
We do not have data connectors. Instead we use Sentinel REST api to upload data and download incidents.

devikamehra replied:
How do you run these calls to Sentinel REST API? Is it using an Azure Function or logic app or something else?
This is useful for us to know since customers will not know how to ingest data if we do not have any data connector.

fs-connect replied:
We have an application (which is available only to customers) that makes use of Sentinel REST API to upload and download.

devikamehra replied:
@fs-connect Thanks for the clarification. How can other users, apart from your customers ingestion this data into Sentinel? If we have clearly defined steps, can you document it in the data connector format applicable to this product. You can find all the templates here - https://github.com/Azure/Azure-Sentinel/tree/master/DataConnectors/Templates

fs-connect replied:
This is a "customer only" product. So, only customers can ingest data into the custom logs.

devikamehra replied:
Got it.

v-spadarthi commented:
@fs-connect : please address the below comments
 1)Could you please move Logo to the Logos folder(https://github.com/Azure/Azure-Sentinel/tree/master/Logos) and logo extension should be .svg file
 2)provide the solution Input file for your reference(https://github.com/Azure/Azure-Sentinel/tree/master/Tools/Create-Azure-Sentinel-Solution)
3) please follow the folder structure like this(https://github.com/Azure/Azure-Sentinel/tree/master/Solutions/OracleWebLogicServer) to create the solution

fs-connect commented:
> @fs-connect : please address the below comments 1)Could you please move Logo to the Logos folder(https://github.com/Azure/Azure-Sentinel/tree/master/Logos) and logo extension should be .svg file 2)provide the solution Input file for your reference(https://github.com/Azure/Azure-Sentinel/tree/master/Tools/Create-Azure-Sentinel-Solution) 3) please follow the folder structure like this(https://github.com/Azure/Azure-Sentinel/tree/master/Solutions/OracleWebLogicServer) to create the solution

Thanks for review.
1. Uploaded a .svg format logo file to Logos directory
2. Uploaded an input file to Tools/Create-Azure-Sentinel-Solution/input directory
3. Followed the folder structure similar to Solutions/OracleWebLogicServer

devikamehra commented on Solutions/Forescout Host Property Monitor/Playbooks/ReadMe.md:
```diff
@@ -0,0 +1,68 @@
+  -	disposition: Represents the actual action in 2-digit format. For example, 00 for Recheck Host, 10 for HTTP Notification.  
+  -	endpoint_ip: Indicates the IP address of the endpoint, which can be an IPv4 or IPv6 address.  
+  -	em_ip: Indicates the IP address of the Enterprise Manager to which the endpoint belongs.  
+  -	info: Indicates addistional information   
```
Typo error - **addistional**

devikamehra commented on Solutions/Forescout Host Property Monitor/Playbooks/ReadMe.md:
```diff
@@ -0,0 +1,68 @@
+action_group="restrict";disposition="31";endpoint_ip=["10.16.141.130"];em_ip=["10.16.147.96"];info=Generated_by_Dnsniff_event  
+Each action has a unique disposition value. The action name, disposition value, and action group supported by eyeExtend for Microsoft Sentinel are as follows:  
+
+Recheck Host	00	manage  
```
Can we represent this information in form of a table. This would make the data more readable.

devikamehra commented on Solutions/Forescout Host Property Monitor/Playbooks/ReadMe.md:
```diff
@@ -0,0 +1,68 @@
+<a name="playbooks">
+
+# Playbook Templates
+This package includes:  
```
Please add deployment buttons of the playbook. Take reference from the example stated earlier.

devikamehra commented on Solutions/Forescout Host Property Monitor/Playbooks/ReadMe.md:
```diff
@@ -0,0 +1,68 @@
+
+<a name="limitations">
+  
+# Known issues and limitations  
```
Since this section does not have any information, we can remove it.

fs-connect replied:
@devikamehra  Thanks for the review. Updated files according to review comments. Thanks.

v-spadarthi commented:
@fs-connect : Could you please keep the solution input file in Data folder under solution folder.

fs-connect commented:
> @fs-connect : Could you please keep the solution input file in Data folder under solution folder.

Thanks for comments. 
Done.

v-mchatla commented:
Hi @@fs-connect 
Can you please incorporate the changes requested by @devikamehra 
Thanks

fs-connect commented:
> Hi @@fs-connect Can you please incorporate the changes requested by @devikamehra Thanks

Hi, 
I have incorporated the changes requested by @devikamehra 
Thank you.

NikTripathi commented:
@v-mchatla Please have a look.

v-mchatla commented:
Hi @fs-connect 
We are working on it, will provide update very soon.
Thanks

shainw commented on Solutions/ForescoutHostPropertyMonitor/Analytic Rules/ForeScout-DNSSniffEventMonitor.yaml:
```diff
@@ -0,0 +1,30 @@
+name: Forescout-DNS_Sniff_Event_Monitor
+description: 'This rule creates an incident when more than certain number of Dnsniff events are generated from a host'
+severity: Medium
+requiredDataConnectors: []
```
@rushriva and @NikTripathi  - Shouldn't this have a required data connector as part of a Solution?

fs-connect replied:
This solution makes use of Sentinel REST API to upload and download data to Log Analytics. Because of this reason, a Data Connector is not needed for solution.

shainw commented on Solutions/ForescoutHostPropertyMonitor/Analytic Rules/ForeScout-DNSSniffEventMonitor.yaml:
```diff
@@ -0,0 +1,30 @@
+relevantTechniques: []
+query:
+  ForescoutHostProperties_CL
+  | where HostProperties_DnsniffEvent_s <> ''
```
I think[ `isnotempty`](https://docs.microsoft.com/en-us/azure/data-explorer/kusto/query/isnotemptyfunction) is what you are looking for here?

fs-connect replied:
Updated to use isnotempty()

shainw commented on Solutions/ForescoutHostPropertyMonitor/Analytic Rules/ForeScout-DNSSniffEventMonitor.yaml:
```diff
@@ -0,0 +1,30 @@
+query:
+  ForescoutHostProperties_CL
+  | where HostProperties_DnsniffEvent_s <> ''
+  | where HostProperties_DnsniffEvent_s matches regex "DNS Query Type:.A;DNS Query/Response:.Query"
```
If there is a regex match on `HostProperties_DnsniffEvent_s`, then why the check above for an empty string?

fs-connect replied:
Removed redundancy check on empty string.

shainw commented on Solutions/ForescoutHostPropertyMonitor/Analytic Rules/ForeScout-DNSSniffEventMonitor.yaml:
```diff
@@ -0,0 +1,30 @@
+  ForescoutHostProperties_CL
+  | where HostProperties_DnsniffEvent_s <> ''
+  | where HostProperties_DnsniffEvent_s matches regex "DNS Query Type:.A;DNS Query/Response:.Query"
+  | extend ipaddress = iif(HostProperties_Ipv4Addr_s <> '', HostProperties_Ipv4Addr_s, (iif(HostProperties_Ipv6Addr_s <> '', HostProperties_Ipv6Addr_s, HostProperties_IpAddr_s)))
```
Again, `isnotempty`?

fs-connect replied:
Updated to use isnotempty()

fs-connect commented:
Hi,
I have addressed all comments from @shainw 
We request to proceed in process. Thanks!

shainw reviewed (commented):
Approved from the Analytic Rule perspective, it is not clear to me if the full solution is approved.  @devikamehra, @rushriva, @v-spadarthi and @v-mchatla anything else needed from your side?

v-spadarthi commented:
Hi @fs-connect
We are working on it, will provide update very soon.
Thanks

v-spadarthi commented:
We are waiting for the update from Eric

v-spadarthi commented:
The update from Eric has not yet been received, so I am checking on it.

fs-connect commented:
@v-spadarthi  Thank you for the update!

v-spadarthi commented:
Still, we are waiting for the update from Eric.

v-mchatla commented:
Hi @v-spadarthi 
Can you please provide update on this.
Thanks

fs-connect commented:
@v-spadarthi , @v-mchatla 
We received feedback from Eric suggesting to add a Data Connector sub-directory. 
We will work on it and submit changes. Thanks

v-spadarthi commented:
@fs-connect :  Data connector we could see below  please fix below
<img width="881" alt="image" src="https://user-images.githubusercontent.com/101796244/194864373-fd5def74-e7ea-4ccf-8f2c-8b4c939fdc13.png">
Please change ispreview =false
<img width="146" alt="image" src="https://user-images.githubusercontent.com/101796244/194865151-b1ae70a0-7f3b-421a-9d80-3943dd1f8d4e.png">
Could you please create a package 2.0.0 by using tool V2 and change the version in solution input file version to 2.0.0.
Create UI definition tested 
<img width="400" alt="image" src="https://user-images.githubusercontent.com/101796244/194864042-4494dc5d-85c6-4134-a431-3170790cea5f.png">

fs-connect commented:
@v-spadarthi 
Thank you for your comments.
We incorporated comments on DataConnector.
But package build using V2 script is giving 3 validation errors. So, we have not updated Package. 
Would previous version of Package (V1) be sufficient? 
Thank you.

v-spadarthi commented:
@fs-connect : Thanks for incorporating the changes in data connector and re-tested it is looking good.
Would previous version of Package (V1) be sufficient?  --No 
Please create a package 2.0.0 by using tool V2 and share the errors we will check. Thanks

fs-connect commented:
@v-spadarthi 
Thanks for the retest and suggestions.
Updated package 2.0.0 using V2.
Following are the validation errors generated when V2 script (.\Tools\Create-Azure-Sentinel-Solution\V2\createSolutionV2.ps1) was executed:
    [-] Controls In Outputs Must Exist (3 ms)
        Count not find element 'playbook1' in .parameters.steps.playbooks Line: 1, Column: 1
    [-] IDs Should Be Derived From ResourceIDs (29 ms)
        .variables._connection-2 not found
        .variables._connector1-source not found
    [-] Variables Must Be Referenced (7 ms)
        Unreferenced variable: _analyticRulecontentId1
        Unreferenced variable: playbook1-MicrosoftSentinelConnectionName

Thank you!

v-spadarthi commented:
Create UI definition 
<img width="387" alt="image" src="https://user-images.githubusercontent.com/101796244/195789959-24c4d617-8df5-4e7e-9603-e1f44ef0d409.png">
Main Template
<img width="929" alt="image" src="https://user-images.githubusercontent.com/101796244/195790622-41868271-cfa2-43a8-8a66-29e5d5259b87.png">
Arm-ttk
<img width="617" alt="image" src="https://user-images.githubusercontent.com/101796244/195791006-b49ed9e5-047a-430c-b60b-ccff81f61635.png"> 
all are fine

v-spadarthi commented:
We are checking with PM team

v-spadarthi commented:
waiting for the update from PM team.

v-spadarthi commented:
@fs-connect : Could you please let us know whether this solution will be published from your end or will it be published by us in our Partner Center.
If the solution is published from your Partner Center, the tier mentioned in solution metadata should be "Partner".

fs-connect commented:
Updated tier to "Partner" and resubmitted Solution metadata and Package.
Thanks!

fs-connect commented:
@v-spadarthi 
From the meeting we had with Eric/Sherry, our understanding is review team would do the merge of PR.
Please let us know if anything need to be done from our side.
Thank you!

v-spadarthi commented:
@fs-connect : Thanks for the changes
While testing we got untitled name for playbooks, please fix it
<img width="501" alt="image" src="https://user-images.githubusercontent.com/101796244/197963241-cb68e956-9cf8-46e0-af69-3fe61b9399f7.png">
Use below link to declare the properties it will resolve above problem
https://github.com/Azure/Azure-Sentinel/tree/master/docs/New%20Playbooks%20Contribution%20Guide#add-metadata

fs-connect commented:
@v-spadarthi : Thanks for the retest and instructions.
Added metadata to the playbook. Please let us know if any errors. Thank you!

v-spadarthi commented:
<img width="728" alt="image" src="https://user-images.githubusercontent.com/101796244/198275350-bfb54be9-03c4-4c5f-859f-ac846d2ad269.png">
<img width="494" alt="image" src="https://user-images.githubusercontent.com/101796244/198275974-657e4f52-084e-478b-b263-2eb1d116049e.png">

<img width="735" alt="image" src="https://user-images.githubusercontent.com/101796244/198275861-e4b72c49-1109-41c8-849e-526a600472b1.png">

Arm-ttk also running fine
<img width="359" alt="image" src="https://user-images.githubusercontent.com/101796244/198276645-b7f22d25-4e24-4bdf-9443-524eb0c69f16.png">

Looking good.

aprakash13 reviewed (approved):
Thanks @fs-connect

v-spadarthi commented:
@fs-connect : Could you please remove the dataconnector file name from  template_ForescoutHostPropertyMonitor.json to template_ and keep it ForescoutHostPropertyMonitor.json and modify the Input file as well. Thanks!

fs-connect commented:
@v-spadarthi : As suggested, changed file name from template_ForescoutHostPropertyMonitor.json to ForescoutHostPropertyMonitor.json. Updated input file. Thanks.

fs-connect commented:
@v-spadarthi and all:
Thank you all for the great help you provided with reviewing, suggesting fixes and merging. Appreciate it.
